packages/shopping-cart/src/state-based-actions.ts

```diff
@@ -0,0 +1,51 @@
+import debugFactory from 'debug';
+import type { ShoppingCartState, ShoppingCartReducerDispatch, CartSyncManager } from './types';
+
+const debug = debugFactory( 'shopping-cart:state-based-actions' );
+
+export function prepareFreshCartForInitialFetch(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	syncManager: CartSyncManager
+): void {
+	const { cacheStatus } = state;
+	if ( cacheStatus === 'fresh' ) {
+		debug( 'triggering fetch of initial cart' );
+		dispatch( { type: 'FETCH_INITIAL_RESPONSE_CART' } );
+		syncManager.fetchInitialCartFromServer( dispatch );
+	}
+}
+
+function prepareInvalidCartForSync(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	syncManager: CartSyncManager
+): void {
+	const { queuedActions, cacheStatus } = state;
+	if ( queuedActions.length === 0 && cacheStatus === 'invalid' ) {
+		debug( 'triggering sync of cart to server' );
+		dispatch( { type: 'REQUEST_UPDATED_RESPONSE_CART' } );
+		syncManager.syncPendingCartToServer( state, dispatch );
+	}
+}
+
+export function createTakeActionsBasedOnState(
+	syncManager: CartSyncManager
+): ( state: ShoppingCartState, dispatch: ShoppingCartReducerDispatch ) => void {
+	let lastState: ShoppingCartState;
+	const takeActionsBasedOnState = (
+		state: ShoppingCartState,
+		dispatch: ShoppingCartReducerDispatch
+	) => {
+		if ( state === lastState ) {
+			return;
+		}
+		lastState = state;
```

Comment:
The `state === lastState` check can be removed. The `dispatch` function already does it and doesn't call `takeActions` if the state hasn't changed. And `fetchInitialCart` is mostly called in a situation where `lastState` is still `undefined`.